### PR TITLE
feat(sandbox): expose services during creation

### LINKED ISSUE #3402: feat: expose sandbox services during creation
## User Story

As an OpenShell user launching a network service in a sandbox, I want to expose that service as part of sandbox creation, so that clients can connect without a separate post-create command or local SSH port-forwarding process.

## Problem Statement

`openshell sandbox create` cannot currently declare a service endpoint. Users must wait for the sandbox to become ready and then run a separate service-exposure operation, or use local port forwarding. SDK callers likewise cannot express service exposure in their sandbox create request.

This makes service-oriented workloads harder to automate and creates a race between sandbox readiness and endpoint registration. It also complicates examples such as launching a Codex app server in an OpenShell sandbox and connecting a local Codex client to it.

## Impact / Why This Matters

The current multi-step workflow requires extra lifecycle coordination, additional error handling, and knowledge of a second API. Local port forwarding also depends on a long-running client-side process and is not suitable when the gateway's routed service endpoint is the desired access path.

A create-time exposure declaration makes service workloads reproducible across the CLI and SDKs, eliminates avoidable orchestration, and gives users one stable endpoint workflow.

## Proposed Design

Allow users to request one or more sandbox service exposures during sandbox creation.

The CLI should support an unnamed exposure with a workflow such as:

```shell
openshell sandbox create \
  --name codex-app-server \
  --expose 4500 \
  -- codex app-server --listen ws://0.0.0.0:4500
```

After the sandbox reaches `Ready`, the CLI should print the routed service URL. Requesting an exposure should keep the sandbox alive after its main command starts and should be incompatible with options that require immediate sandbox cleanup.

The public sandbox create API and the Rust, Python, TypeScript, and Go SDK create options should accept service exposure declarations containing a target port and an optional service name. Named exposures should use the same naming and routing behavior as the existing service-exposure command.

The routed service boundary should preserve the existing behavior of consuming gateway authentication and not forwarding the gateway `Authorization` header to the sandbox application. Application-level authentication should use a distinct credential mechanism.

Add an example under `examples/` that runs the Codex app server inside an OpenShell sandbox and shows how a local Codex client connects through the exposed WebSocket URL.

## Acceptance Criteria

- [ ] `openshell sandbox create --expose <PORT>` requests an unnamed routed service endpoint as part of sandbox creation.
- [ ] The CLI prints the service URL after the sandbox reaches `Ready`.
- [ ] Create-time exposure keeps the sandbox and rejects conflicting immediate-cleanup options.
- [ ] The create API supports one or more exposures with a target port and optional service name.
- [ ] Invalid ports, invalid service names, duplicate service names, and excessive exposure counts are rejected before sandbox creation is persisted.
- [ ] Rust, Python, TypeScript, and Go SDK callers can supply exposures in direct and template-based sandbox create operations.
- [ ] Existing standalone service exposure behavior remains available and compatible.
- [ ] Gateway authentication headers are not forwarded to the exposed sandbox service.
- [ ] Documentation includes CLI and SDK examples.
- [ ] An `examples/` workflow launches a Codex app server in a sandbox and connects a local Codex client through the routed WebSocket endpoint.
- [ ] Unit and integration tests cover API validation, endpoint registration, CLI request construction, and SDK request mapping.

## Alternatives Considered

Continue requiring `openshell sandbox service expose` after creation. This preserves the existing API but forces every caller to coordinate readiness and perform a second mutation.

Use `openshell sandbox create --forward` exclusively. Port forwarding remains useful for local-only access, but it requires a persistent client-side forwarding process and does not provide a gateway-routed service endpoint.

Forward the gateway `Authorization` header to the workload. This would conflate gateway authentication with application authentication and disclose a gateway credential to sandbox code, so the service router should continue stripping it.

## Agent Investigation

The existing gateway service router already supports named and unnamed `ServiceEndpoint` records and WebSocket upgrades. The create path can reuse the same exposure validation and endpoint persistence behavior. The maintained SDK surfaces are Rust, Python, TypeScript, and Go.

### PR DESCRIPTION
**Pull request by drew**
## Summary

Add create-time sandbox service exposure across the gateway API, CLI, and maintained SDKs, returning routed service URLs keyed by service name. Include a Codex app-server example that uses gateway-managed credential refresh and connects a local Codex client through the exposed WebSocket endpoint.

## Related Issue

Closes #3402

## Changes

- Add create-time service exposure requests and service URL maps to the protobuf API
- Support `openshell sandbox create --expose PORT`, structured URL output, validation, and rollback when endpoint persistence fails
- Expose named and unnamed services from the Rust, Python, TypeScript, and Go SDK create APIs
- Normalize create-time URLs against the externally configured gateway port
- Add Codex OAuth refresh metadata so the gateway owns refresh material
- Add the `examples/codex-app-server` image, launcher, and local-client workflow
- Update sandbox documentation, architecture notes, and the public CLI skill

## Testing

- [x] `mise run pre-commit` passes
- [x] Unit tests added/updated
- [ ] E2E tests added/updated (not included; the example was exercised manually against a local gateway)

Additional focused tests:

- `cargo test -p openshell-server create_sandbox_begins_rollback_when_service_exposure_fails`
- `cargo test -p openshell-server create_sandbox_registers_requested_service_exposures`
- `cargo test -p openshell-cli --test sandbox_create_lifecycle_integration sandbox_create_json_stdout_is_parseable`
- `cargo test -p openshell-cli service_url_for_gateway_uses_external_gateway_port`

## Checklist

- [x] Follows [Conventional Commits](https://www.conventionalcommits.org/)
- [x] Commits are signed off (DCO)
- [x] Architecture docs updated (if applicable)

### CODE REVIEW

**Comment by mrunalp** [comment]
### Potential Concerns

  1. Create replay would lose service_urls. Current main reconstructs replayed CreateSandbox responses using only the sandbox ID. After this PR, a retry with the same request_id could successfully replay creation but return an empty URL map. The replay receipt/restore logic needs to
     preserve or reconstruct the URLs, with a retry test. Replay logic (https://github.com/NVIDIA/OpenShell/blob/c108c31696bee299a374cda0b2969cf9ffe4e74d/crates/openshell-server/src/grpc/mutation_replay/ordinary.rs#L372-L406)

  2. Rollback can delete a same-name replacement. On exposure failure, the handler knows the created sandbox ID but calls deletion by name. If concurrent deletion and recreation occur before name resolution, rollback can target the replacement. Use an ID-fenced internal deletion
     path. Rollback call (https://github.com/NVIDIA/OpenShell/blob/97fddfcc3211816e56a7b77a3892120b3d1577cf/crates/openshell-server/src/grpc/sandbox.rs#L545-L580), name resolution
     (https://github.com/NVIDIA/OpenShell/blob/c108c31696bee299a374cda0b2969cf9ffe4e74d/crates/openshell-server/src/compute/mod.rs#L1694-L1729)

  3. The Codex example eventually invokes the placeholder refresh token. It writes auth_mode: "chatgpt" and a fake refresh token. Codex cannot parse the opaque OpenShell access-token handle as a JWT, so after eight days it falls back to proactive refresh using that placeholder,
     causing recurring refresh failures and latency. Use Codex’s externally managed chatgptAuthTokens mode instead. Example launcher (https://github.com/NVIDIA/OpenShell/blob/97fddfcc3211816e56a7b77a3892120b3d1577cf/examples/codex-app-server/start-codex-app-server#L43-L55), Codex
     refresh logic (https://github.com/openai/codex/blob/rust-v0.155.0/codex-rs/login/src/auth/manager.rs#L2959-L2980), external auth mode (https://github.com/openai/codex/blob/rust-v0.155.0/codex-rs/protocol/src/auth.rs#L7-L17)

**Comment by drew** [comment]
Addressed all three review points and rebased onto current `main`:

- create mutation replay now preserves `service_urls`, with a retry/replay regression test
- failed exposure rollback now deletes by the created sandbox ID, so a same-name replacement cannot be removed; added a replacement-race regression test
- the Codex launcher now uses `chatgptAuthTokens` external auth mode with no synthetic refresh token, leaving refresh ownership with the gateway

Validation: pre-commit, full `openshell-server` suite, sandbox create lifecycle integration (40 passed, 2 ignored), TypeScript SDK CI, Python tests, and focused Go SDK tests.

File: providers/codex.yaml
Comment (drew) [inline providers/codex.yaml:1]: @johntmyers do you mind taking a look at this change and making sure it looks ok?

File: examples/codex-app-server/README.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+## 2. Create the provider
+
+These commands create a provider from the current host login, move the refresh
+token into gateway-only refresh material, and rotate the access token once. The
+sandbox receives opaque handles for only the access token and account ID; it
+never receives the refresh token. Run these commands once per gateway.
+
+```shell
+openshell provider create \
```
Comment (drew) [inline examples/codex-app-server/README.md:53]: > **gator-agent**

**Warning — GATOR-207dcb6b-01 · Import the Codex profile before creating the provider**

**Summary:** An operator following this example on a fresh default gateway reaches `provider create --type codex` with an empty provider-profile catalog. The exact profile lookup then fails with `provider profile 'codex' not found`, so credential refresh and sandbox creation never run. The profile file itself says it is not loaded automatically.

**Fix:** Before `provider create`, add a lint/import step for `../../providers/codex.yaml`, or explicitly require and verify an already-imported `codex` profile. Cover the clean-gateway path in the example smoke coverage.

**Verify:** Start a default gateway with no imported profiles and follow the example from its directory; provider creation, refresh configuration, rotation, and sandbox creation must proceed without a missing-profile error.

<details>
<summary>Agent context</summary>

- **Location:** `examples/codex-app-server/README.md:40`
- **Ownership:** This PR adds the example and modifies the explicitly imported `providers/codex.yaml` profile without adding an automatic catalog source.

</details>

**Review by drew — COMMENTED** [review]
> **gator-agent**

## PR Review Status

Thanks @drew. I checked your update against the three concerns you acknowledged: the current head preserves service URLs across mutation replay, fences rollback by the created sandbox ID, and configures Codex with externally managed `chatgptAuthTokens`. Those items are resolved, but the new example still cannot run on the documented fresh-gateway path because it never imports the Codex provider profile.

Action required: add the missing profile-import or explicit pre-import verification step to the Codex example, with clean-gateway smoke coverage.

Blocking findings:
- `GATOR-207dcb6b-01`: the Codex example uses a provider profile that a default gateway does not load automatically.

Carried findings:
- None

<details>
<summary>Gator metadata</summary>

- Validation: Maintainer-authored implementation of linked issue #3402 is project-valid.
- Docs: User-facing docs are present, but the new Codex example omits a required provider-profile setup step.
- Checks: Current-head Branch Checks, Helm Lint, and Trivy Changes are green.
- E2E: `test:e2e` is required for sandbox lifecycle and gateway behavior; dispatch will be decided after review feedback is resolved.
- Head SHA: `207dcb6b9f58aa4cbacd4031986a2d6a2c97f89e`
- Base SHA: `fa0bfa490e42c87a74a70be6ebb40faee7fb8faa`
- Merge base SHA: `fa0bfa490e42c87a74a70be6ebb40faee7fb8faa`
- Patch ID: `7e991d7aa0b3c5bdacdab8fb4a09f4053a85d505`
- Gator payload: `9`
- Review mode: `initial`
- Previous reviewed SHA: none
- Review budget exhausted: no
- Maintainer decision required: no
- Next state: `gator:in-review`

</details>

**Review by drew — COMMENTED** [review]
> **gator-agent**

## PR Review Status

Thanks @drew. I checked the latest commit against the open provider-profile finding. Bundling `examples/codex-app-server/codex.yaml` and documenting lint/import before provider creation fixes the manual clean-gateway path, but the requested executable smoke coverage is still absent, so the finding is only partially resolved.

Action required: add a constrained smoke test or example smoke script that starts with an empty workspace profile catalog, lints and imports the bundled profile, and verifies provider creation succeeds before refresh and sandbox creation.

Blocking findings:
- No new blocking findings.

Carried findings:
- `GATOR-207dcb6b-01`: the documented profile import is fixed, but no repository smoke path exercises the clean-gateway setup through provider creation.

<details>
<summary>Gator metadata</summary>

- Validation: Maintainer-authored implementation of linked issue #3402 is project-valid.
- Docs: The Codex example now documents the required profile import; clean-gateway smoke coverage remains missing.
- Checks: Current-head required Branch Checks are still running; Helm Lint and Trivy Changes are green.
- E2E: `test:e2e` is required for sandbox lifecycle, gateway behavior, and provider credential flow; dispatch will be decided after review feedback is resolved.
- Head SHA: `aa7d9bced3c4721d09872d00454f873738099143`
- Base SHA: `fa0bfa490e42c87a74a70be6ebb40faee7fb8faa`
- Merge base SHA: `fa0bfa490e42c87a74a70be6ebb40faee7fb8faa`
- Patch ID: `cb5a7423fecd0049c71d4c3a367d16d9ab31e468`
- Gator payload: `9`
- Review mode: `follow_up`
- Previous reviewed SHA: `207dcb6b9f58aa4cbacd4031986a2d6a2c97f89e`
- Review budget exhausted: no
- Maintainer decision required: no
- Next state: `gator:in-review`

</details>

File: providers/codex.yaml
Comment (drew) [reply providers/codex.yaml:1]: Wound up reverting this and just putting a codex profile in the example instead.

File: examples/codex-app-server/README.md (hunk 1 above)
Comment (drew) [reply examples/codex-app-server/README.md:53]: The manual clean-gateway path is now fixed: the example bundles, lints, and imports its provider profile before provider creation. We are waiving dedicated smoke coverage for this finding because repository examples are not currently required to have individual CI tests. The existing feature and provider test coverage remains unchanged.

**Comment by drew** [comment]
Maintainer disposition for `GATOR-207dcb6b-01`: the functional finding is resolved. The Codex example now bundles, lints, and imports its provider profile before provider creation, so the documented clean-gateway path is complete.

We are waiving dedicated example smoke coverage for this PR because repository examples are not currently required to have individual CI tests. No further change is requested for this finding.

**Comment by drew** [comment]
> **gator-agent**

## Blocked

Thanks @drew. I verified your maintainer disposition for `GATOR-207dcb6b-01`: the bundled profile now completes the clean-gateway setup path, and your waiver of dedicated example smoke coverage resolves the remaining test request. The current head is rebase-equivalent to the reviewed patch, so no additional code review is needed.

The current E2E run failed in `rust-docker` and rootless Podman conformance, although both jobs passed on the immediately preceding run of the same effective patch. Branch Checks are also still running.

Action required: a maintainer should re-run the failed jobs in Branch E2E Checks run `35495545168`; if either failure repeats, inspect its logs and determine whether an author change is needed.

<details>
<summary>Gator metadata</summary>

- Head SHA: `8733df2c8db75379b512062df0129f96937721b3`
- Base SHA: `29e89a2f2289ad538c195e136baaaac2a92a3a2e`
- Patch ID: `cb5a7423fecd0049c71d4c3a367d16d9ab31e468`
- Gator payload: `9`
- Review mode: `already_reviewed`
- Checks: E2E failed; Branch Checks pending; Helm Lint and Trivy Changes green.
- E2E: Maintainer rerun required.
- Next state: `gator:blocked`
- Blocked reason: `ci_rerun_required`

</details>

**Review by johntmyers — DISMISSED** [review]

**Review by drew — COMMENTED** [review]
> **gator-agent**

## PR Review Status

Thanks @drew. I reviewed the latest author-only change that allows the npm-installed Codex binary under `/usr/local/lib/node_modules`, and it introduces no new blocking findings. Your earlier maintainer disposition remains intact: the clean-gateway profile setup is fixed, and the dedicated example smoke-test request was explicitly waived. Required Branch Checks and E2E are now running on this head.

Blocking findings:
- No blocking findings remain

Carried findings:
- None; `GATOR-207dcb6b-01` is resolved and waived

<details>
<summary>Gator metadata</summary>

- Validation: Maintainer-authored implementation of linked issue #3402 is project-valid.
- Docs: User-facing docs and the Codex example are present; this follow-up profile-path change needs no additional Fern update.
- Checks: Current-head Branch Checks are running; Helm Lint and Trivy Changes are green.
- E2E: `test:e2e` is applied and current-head E2E jobs are queued or running.
- Head SHA: `a6e7bf425f657c4dd5cbc41411e0412449acc6d3`
- Base SHA: `29e89a2f2289ad538c195e136baaaac2a92a3a2e`
- Merge base SHA: `29e89a2f2289ad538c195e136baaaac2a92a3a2e`
- Patch ID: `6d245ccefc320f34a736ae7efef347f25909efec`
- Gator payload: `9`
- Review mode: `follow_up`
- Previous reviewed SHA: `aa7d9bced3c4721d09872d00454f873738099143`
- Review budget exhausted: no
- Maintainer decision required: no
- Next state: `gator:watch-pipeline`

</details>

**Comment by drew** [comment]
> **gator-agent**

## Monitoring Complete

Monitoring is complete because this PR has merged.

Final status: the current-head required checks, including E2E, passed before merge; the PR was awaiting maintainer approval when it reached its terminal state.

I removed the active `gator:*` label because there is nothing left for gator to monitor on this PR.

<details>
<summary>Gator metadata</summary>

- Head SHA: `a6e7bf425f657c4dd5cbc41411e0412449acc6d3`
- Gator payload: `9`
- Final state: merged

</details>
